src/tls/cockpit-certificate-ensure.c

```diff
@@ -60,62 +120,210 @@ check_expiry (const char *filename,
+
+  return creds;
+}
+
 static bool
-have_certificate (bool verbose)
+cockpit_certificate_find (CertificateKeyPair *result,
+                          bool                verbose)
 {
   char *error = NULL;
-  char *filename = cockpit_certificate_locate (true, &error);
+  char *certificate_filename = cockpit_certificate_locate (true, &error);
 
   if (error != NULL)
     errx (EXIT_FAILURE, "%s", error);
 
-  if (filename == NULL)
+  if (certificate_filename == NULL)
     {
-      debug (ENSURE, "Couldn't locate any certificate");
-
       if (verbose)
-        printf ("No certificate found\n");
+        printf ("Unable to find any certificate file\n");
 
       return false;
     }
 
-  if (strstr (filename, "/0-self-signed.cert"))
+  certificate_and_key_read (result, certificate_filename);
+
+  gnutls_certificate_credentials_t creds = certificate_and_key_parse_to_creds (result);
+
+  bool expired = is_selfsigned (certificate_filename) && check_expiry (creds);
+  if (expired)
     {
-      debug (ENSURE, "Certificate is self-signed, checking expiry");
-      if (!check_expiry (filename, verbose))
-        return false;
-    }
+      if (verbose)
+        printf ("Found self-signed %s, but it needs to be reissued\n",
+                result->filename_for_errors);
 
-  debug (ENSURE, "Certificate looks good: %s", filename);
+      certificate_and_key_clear (result);
+    }
 
-  if (verbose)
-    printf ("Would use certificate: %s\n", filename);
+  gnutls_certificate_free_credentials (creds);
+  free (certificate_filename);
 
-  return true;
+  return !expired;
 }
 
-#define COCKPIT_CERTIFICATE_HELPER   LIBEXECDIR "/cockpit-certificate-helper"
+static void
+cockpit_certificate_selfsign (CertificateKeyPair *result)
+{
+  pid_t pid;
+  int r = posix_spawn (&pid, COCKPIT_CERTIFICATE_HELPER, NULL, NULL,
+                       (char *[]){ COCKPIT_CERTIFICATE_HELPER, "selfsign", NULL },
+                       NULL);
+
+  if (r != 0)
+    errx (EXIT_FAILURE, "posix_spawn: %s: %s",
+          COCKPIT_CERTIFICATE_HELPER, strerror (r));
+
+  int status;
+  do
+    r = waitpid (pid, &status, 0);
+  while (r == -1 && errno == EINTR);
+
+  if (r < 0)
+    err (EXIT_FAILURE, "wait: %s", COCKPIT_CERTIFICATE_HELPER);
+
+  if (!WIFEXITED (status) || WEXITSTATUS (status) != 0)
+    errx (EXIT_FAILURE, "%s exited with non-zero status %d",
+          COCKPIT_CERTIFICATE_HELPER, WEXITSTATUS (status));
+
+  certificate_and_key_read (result, COCKPIT_SELFSIGNED_FILENAME);
+
+  /* We just generated this ourselves, so we don't bother to check it
+   * for validity.
+   */
+}
 
 int
 main (int argc, char **argv)
 {
+  CertificateKeyPair result = { };
   bool check = false;
 
-  if (argc == 1)
+  const char *runtime_directory = getenv ("RUNTIME_DIRECTORY");
+
+  /* We need RUNTIME_DIRECTORY set to actually do anything, so if it's
+   * not set then --check is mandatory.
+   */
+  if (argc == 1 && runtime_directory)
```

Comment:
I think that's the change that breaks `podman container runlabel INSTALL cockpit/ws` -- because in this scenario it wouldn't have a `$RUNTIME_DIRECTORY`. It would be good to at least point that out in the error message.